### PR TITLE
fix: Incorrect results when planning aggregate expressions

### LINKED ISSUE #7361: InfluxQL-IOx: Incorrect results for aggregate expression
**Steps to reproduce:**

Using the end-to-end test, add the following query:

```sql
SELECT MEAN(usage_idle) + MEAN(foo) FROM cpu;
```

**Expected output:**

```
-- InfluxQL: SELECT MEAN(usage_idle) + MEAN(foo) FROM cpu;
+------------------+----------------------+--------------------------+
| iox::measurement | time                 | mean_usage_idle_mean_foo |
+------------------+----------------------+--------------------------+
| cpu              | 1970-01-01T00:00:00Z |                          |
```

**Actual output:**

```
-- InfluxQL: SELECT MEAN(usage_idle) + MEAN(foo) FROM cpu;
+------------------+----------------------+--------------------------+
| iox::measurement | time                 | mean_usage_idle_mean_foo |
+------------------+----------------------+--------------------------+
| cpu              | 1970-01-01T00:00:00Z |                          |
| cpu              | 1970-01-01T00:00:00Z |                          |
| cpu              | 1970-01-01T00:00:00Z |                          |
| cpu              | 1970-01-01T00:00:00Z |                          |
| cpu              | 1970-01-01T00:00:00Z |                          |
| cpu              | 1970-01-01T00:00:00Z |                          |
+------------------+----------------------+--------------------------+
```

**Other information:**

The following query produces incorrect results for a similar reason:

```sql
SELECT MEAN(usage_idle) + MEAN(bytes_free) FROM cpu, disk;
```

### PR DESCRIPTION
**Pull request by stuartcarnie**
Closes #7361

This PR ensures that the correct number of rows are returned when an aggregate expression contains a mix of fields that exist and do not exist in an single measurement being queried.

### CODE REVIEW

File: iox_query_influxql/src/plan/planner.rs [hunk 1]
```diff
@@ -453,9 +457,6 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
         //
         // [SUM(foo), COUNT(foo)]
         let aggr_exprs = find_aggregate_exprs(&select_exprs);
```
Comment (wolffcm) [inline iox_query_influxql/src/plan/planner.rs:459]: I think that this will return a non-empty vector for the test cases relevant to this fix.

File: iox_query_influxql/src/plan/planner.rs [hunk 2]
```diff
@@ -483,6 +484,14 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
             vec![]
         };
 
+        if aggr_exprs.is_empty() && aggr_group_by_exprs.is_empty() {
+            // All the field expressions were reduced to `NULL` and there is no group by,
+            // therefore, the result set is a single row.
+            //
+            // This is required for InfluxQL compatibility.
+            return Ok((LogicalPlanBuilder::empty(true).build()?, select_exprs));
```
Comment (wolffcm) [inline iox_query_influxql/src/plan/planner.rs:492]: I think that this must be the path taken for the fix to have the desired effect. But I don't see how this path gets taken since I think that `aggr_exprs.is_empty()` would be false. I must be confused somewhere, can you clarify?

File: iox_query_influxql/src/plan/planner.rs [hunk 3]
```diff
@@ -439,6 +439,10 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
         group_by_tag_set: &[&str],
         schemas: &Schemas,
     ) -> Result<(LogicalPlan, Vec<Expr>)> {
+        if !ctx.is_aggregate {
+            return Ok((input, select_exprs));
+        }
```
[the remark this answers is not in the document]
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner.rs:444]: I can completely understand your confusion, as the same has occurred for me numerous times as I continue to read and unpack the original source. A lot of this has to do with the fact that InfluxQL and the TSM data model is _schema-less_ and that InfluxQL makes every effort to execute your query without error.


## TL;DR

After rewriting the `select_exprs`, it is possible that aggregate functions have been replaced with `NULL` for columns that do not exists in the table, hence the `find_aggregate_exprs` could return an empty vector.


## Explanation

The `is_aggregate` flag indicates the current `SELECT` query has aggregate functions, such as `COUNT` or `SUM`. If that is the case, it has implications on how the query is processed and the number of results returned (such as returning a single row per group), and therefore we must proceed with processing the query as an aggregate to ensure these properties are upheld. 

Your first question:

> I think that this will return a non-empty vector for the test cases relevant to this fix.
>
> https://github.com/influxdata/influxdb_iox/blob/839816abf1b8c6e7f7d714e7821cd066d0ed2b1e/iox_query_influxql/src/plan/planner.rs#L455

is a reasonable assumption, however, because we have to deal with non-existent fields whilst processing the current table, we replace these references with `NULL` values:

https://github.com/influxdata/influxdb_iox/blob/839816abf1b8c6e7f7d714e7821cd066d0ed2b1e/iox_query_influxql/src/plan/planner.rs#L803

and then we reduce field expressions that have `NULL` values to `NULL`:

https://github.com/influxdata/influxdb_iox/blob/839816abf1b8c6e7f7d714e7821cd066d0ed2b1e/iox_query_influxql/src/plan/planner.rs#LL659C30-L659C30

via this function:

https://github.com/influxdata/influxdb_iox/blob/839816abf1b8c6e7f7d714e7821cd066d0ed2b1e/iox_query_influxql/src/plan/planner_rewrite_expression.rs#L151-L156

So, more concretely, say we have:

```sql
SELECT MEAN(usage_idle) + MEAN(non_existent) AS col0 FROM cpu
```

In SQL, this would simply fail. However, InfluxQL will execute it. The first pass to convert the InfluxQL expression `MEAN(usage_idle) + MEAN(non_existent) AS col0` to a DataFusion logical expression must validate that all column references exist in the current table. Any column references that don't exist are replaces with a `NULL` scalar:

```
MEAN(usage_idle) + MEAN(NULL) AS col0
```

and then expressions that contain `NULL`s are reduced to `NULL`, thus rewriting evaluates to

```
NULL AS col0
```

With that knowledge, we know the following line will return an empty vector:

```
https://github.com/influxdata/influxdb_iox/blob/839816abf1b8c6e7f7d714e7821cd066d0ed2b1e/iox_query_influxql/src/plan/planner.rs#L455
```

With that in mind, we now have this check:

https://github.com/influxdata/influxdb_iox/blob/9ed6696150d5a9bde8f77a883919fa2db74e6bb4/iox_query_influxql/src/plan/planner.rs#L487-L493

If there are no aggregates or a group by, then there must be a single row in the result set, so we return an empty plan with a single row.

Otherwise, we continue as the aggregate must have a `GROUP BY TIME` or `GROUP BY tag0[, ...]`. Using the above example again, but adding a `GROUP BY`:

```sql
SELECT MEAN(usage_idle) + MEAN(non_existent) AS col0 FROM cpu GROUP BY host
```

The aggregate is still reduces to `NULL` and thus the `find_aggregate_exprs` vector is empty, however we end up with a logical plan something like the following SQL:

```
WITH t AS (SELECT host FROM cpu GROUP BY 1)
SELECT host, NULL as col0 FROM t
```

File: iox_query_influxql/src/plan/planner.rs (hunk 1 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner.rs:465]: See: https://github.com/influxdata/influxdb_iox/pull/7431#discussion_r1156557803

File: iox_query_influxql/src/plan/planner.rs (hunk 2 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner.rs:499]: See: https://github.com/influxdata/influxdb_iox/pull/7431#discussion_r1156557803

File: iox_query_influxql/src/plan/planner.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (wolffcm) [reply iox_query_influxql/src/plan/planner.rs:444]: Thank you for this thorough explanation! Now that I understand it, it makes a lot of sense, and I can see why you need to store `ctx.is_aggregate` as state to preserve the semantics of the InfluxQL query, and why it might not match up with the call to `find_aggregate_exprs`.
